### DISCUSSION TITLE
I want the user to be able to handle data sending and receiving using the client used in the OTel Java agent. Is this possible?

### DISCUSSION THREAD

**Discussion by zerocooldog** [open-telemetry/opentelemetry-java-instrumentation#12362] [Q&A] [answered]
I understood that the gRPC exporter in the OpenTelemetry Java agent can only send data. Is that correct?
I want the user to be able to handle data sending and receiving using the client used in the OTel Java agent. Is this possible?
Or do I need to implement a separate gRPC client for the user to handle the sending and receiving?
I want to implement a gRPC client directly, but if there is a client used in the OpenTelemetry Java agent, I want to use it to process data

**Comment by laurit** [comment]
> I understood that the gRPC exporter in the OpenTelemetry Java agent can only send data. Is that correct?

correct. You can find the grpc exporter used by the agent in https://github.com/open-telemetry/opentelemetry-java repository.

**Reply by zerocooldog** [reply]
@laurit I saw it but I didn't understand it. What should I look at?? I want to send my own custom data to the grpc server, in addition to the defined exporters like traces, metrics, logs.

Would it be helpful to examine examples of GrpcExporterBuilder or UpstreamGrpcSender? 

This problem is too difficult. I'm not sure about this one either.

```

            new GrpcExporterBuilder<>(
                    "exporter",
                    "type",
                    10,
                    new URI("http://localhost"),
                    () -> ServerStatusServiceGrpc::newFutureStub,
                    "/path")
                    .setChannel(ManagedChannelBuilder.forTarget("localhost").build())
                    .build())
```




I simply want to send the following proto source from the agent to the server.


```

syntax = "proto3";

package opentelemetry.proto.collector.hamonica2.v1;
import "opentelemetry/proto/common/v1/common.proto";

service CommandService {
  rpc Export (CommandRequest) returns (CommandResponse);
}

message CommandRequest{

  string serverType = 1;
  string serviceInstanceId = 2;
  string serviceName = 3;
  string serviceNamespace = 4;
  string command = 5;
  map<string, string>	parameter = 6; 

}

message CommandResponse {
  bytes data = 1; 
  common.v1.ResponseStatus status = 2;
  string message = 3;
}


```

**Reply by laurit** [reply, marked as answer]
If you need to send custom info then you will have to implement it yourself with whatever tools you like.

**Reply by zerocooldog** [reply]
@laurit Got it. Any chance of getting this feature added later on?

**Reply by laurit** [reply]
It is unclear to me what the feature it that you are requesting.

**Reply by zerocooldog** [reply]
@laurit A feature to allow users to send custom data. 

While I believe it may not fully align with OTEL's intended purpose, I wanted to inquire about the complexities involved in implementing a necessary feature from a third-party perspective, given the numerous considerations

**Reply by laurit** [reply]
Otel agent can be extended so users are free to implement sending whatever they like. Since what you are building seems to be a custom signal you could look at how the exporters for other signals are implemented in https://github.com/open-telemetry/opentelemetry-java If you have specific proposals on apis that could assist you then you'll have to raise them in opentelemetry-java as that is where the exporters live. Alternatively you could consider whether you really need a custom signal, perhaps you could reuse the logs signal, log body is an AnyValue.

**Reply by zerocooldog** [reply]
@laurit ok. Thank you.
